src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -95,28 +85,121 @@ public class PublishPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
                 // This should not happen since the build id comes from build environment. But a user may override that so we must be careful.
                 throw new ArgumentException(StringUtil.Loc("BuildIdIsNotValid", buildIdStr));
             }
+            
+            if(artifactType == pipelineType) {
+                string hostType = context.Variables.GetValueOrDefault(WellKnownDistributedTaskVariables.HostType)?.Value; 
+                if (!string.Equals(hostType, "Build", StringComparison.OrdinalIgnoreCase)) {
+                    throw new InvalidOperationException(
+                        StringUtil.Loc("CannotUploadFromCurrentEnvironment", hostType ?? string.Empty)); 
+                }
+
+                if (String.IsNullOrWhiteSpace(artifactName))
+                {
+                    string jobIdentifier = context.Variables.GetValueOrDefault(WellKnownDistributedTaskVariables.JobIdentifier).Value;
+                    var normalizedJobIdentifier = NormalizeJobIdentifier(jobIdentifier);
+                    artifactName = normalizedJobIdentifier;
+                }
+
+                if(!PipelineArtifactPathHelper.IsValidArtifactName(artifactName)) {
+                    throw new ArgumentException(StringUtil.Loc("ArtifactNameIsNotValid", artifactName));
+                }
+
+                string fullPath = Path.GetFullPath(targetPath);
+                bool isFile = File.Exists(fullPath);
+                bool isDir = Directory.Exists(fullPath);
+                if (!isFile && !isDir)
+                {
+                    // if local path is neither file nor folder
+                    throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
+                }
+
+                // Upload to VSTS BlobStore, and associate the artifact with the build.
+                context.Output(StringUtil.Loc("UploadingPipelineArtifact", fullPath, buildId));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.UploadAsync(context, projectId, buildId, artifactName, fullPath, token);
+                context.Output(StringUtil.Loc("UploadArtifactFinished"));
 
-            string fullPath = Path.GetFullPath(targetPath);
-            bool isFile = File.Exists(fullPath);
-            bool isDir = Directory.Exists(fullPath);
-            if (!isFile && !isDir)
-            {
-                // if local path is neither file nor folder
-                throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
             }
+            else if (artifactType == fileShareType)
+            {
+                string fileSharePath = context.GetInput(ArtifactEventProperties.FileSharePath, required: true);
```

Comment:
too much happening in this one function , can we split it into multiple functions, one for publsih pipeline type and one for UNC paths.